[refactor] split constructor/fields types from Types to Data_types

gasche opened the pull request:
One must be careful about changing typing/Types, because (some of) the types it defined are used in .cmi files, so any change to those requires a delicate bootstrap and magic number bump.

This refactoring splits off a part of Types that is *not* used in .cmi files -- information about record fields and variant constructors. They move to a new typing/Data_types compilation unit, which can evolve more freely.

gasche committed:
[refactor] split constructor/fields types from Types to Data_types

One must be careful about changing typing/Types, because (some of) the
types it defines are used in .cmi files, so any change to those
requires a delicate bootstrap and magic number bump.

This refactoring splits off a part of Types that is *not* used in .cmi
files -- information about record fields and variant
constructors. They move to a new typing/Data_types compilation unit,
which can evolve more freely.

Octachron commented:
Should the new module name be `Datatypes` to match `Datarepr`? (or `Datadesc`? `Labeldesc`?)

Overall, I agree that detaching this module from `Types` make sense : 

- the field and constructor descriptions are not types
- they are not used elsewhere in the `Types` modules
- they don't need to be privy to `Types` internal.

I would propose to move the two functions of `Btypes` `label_is_poly` and `cstr_type_path` to the new module to remove the `Btype->Datatypes` dependency.

Similarly, I would rather avoid opening the full module in the mli for `Ctypes` and`Typecore` (and possibly `Datarepr`) because the new module types are only used once or twice in the interface.

gasche committed:
[refactor] Btype.cstr_type_path => Data_types.cstr_res_type_path

Suggested-by: Florian Angeletti <florian.angeletti@inria.fr>

gasche committed:
[refactor] remove Btype.label_is_poly

Suggested-by: Florian Angeletti <florian.angeletti@inria.fr>

gasche commented:
I applied your recommendations and rebased.

Changes:
- Btype.cstr_type_path is now Data_types.cstr_res_type_path
- I removed Btype.label_is_poly (I was not too exited at adding a dependency from Data_types (mostly a types-only module) to Btype (a types-and-logic module))
- I removed a two of the global open you mentioned (remark: reformating .mli lines to avoid 80 columns is not fun)

I prefer Data_types to Datatypes, because I can see a use for other Foo_types module, for example my in-progress PR has a Head_shape_types module, and Headshapetypes would be ugly. We could change Datadescr into something else, in time (never).

Octachron reviewed (approved):
The change looks fine to me.

gasche commented:
Thanks! I will fix the typos, massage the history (removing the opens could go in the first commit), merge, and rebase #13467.
